### PR TITLE
bpo-42955: Add sys.modules_names

### PR DESCRIPTION
**Pull request by vstinner**
Add sys.module_names attribute: the list of the standard library
module names.

https://bugs.python.org/issue42955

### CODE REVIEW

**Comment by vstinner** [comment]
The most important part of the PR is the ``Tools/scripts/generate_module_names.py`` script which generates the list.

It ignores the following modules. I'm not sure if we should ignore them or not.

```
IGNORE = {
    '__pycache__',
    'site-packages',

    # Helper modules for public module
    # (ex: _osx_support is used by sysconfig)
    '_aix_support',
    '_collections_abc',
    '_compat_pickle',
    '_compression',
    '_markupbase',
    '_osx_support',
    '_sitebuiltins',
    '_strptime',
    '_threading_local',
    '_weakrefset',

    # Used to bootstrap setup.py
    '_bootsubprocess',

    # test modules
    'test',
    '__phello__.foo',

    # pure Python implementation
    '_py_abc',
    '_pydecimal',
    '_pyio',
}
```

I chose to ignore these modules to make the list looking nicer. But it makes the list "not correct".

For Windows, I'm lazy and hardcoded the list since it's short and is no updated often:

```
WINDOWS_MODULES = (
    "_msi",
    "_testconsole",
    "msvcrt",
    "winreg",
    "winsound"
)
```

**Comment by vstinner** [comment]
I updated the documentation:

> Some special stdlib modules are excluded from this list, like "test" and private helper modules of public modules.

I also fixed the code to also list sub-packages: the new module count is now 313.

Should we also list package sub-modules like asyncio.base_events?

**Comment by vstinner** [comment]
ensurepip._bundled is not included since it doesn't contain any .py file, only .whl files.

**Comment by vstinner** [comment]
> ensurepip._bundled is not included since it doesn't contain any .py file, only .whl files.

My bad, it's listed: there is ``Lib/ensurepip/_bundled/__init__.py``.

**Comment by vstinner** [comment]
@pablogsal @serhiy-storchaka: Ok, this PR is now ready for your review :-) I updated the documentation to explicit which modules are included and which are excluded:

```
A tuple of strings giving the names of standard library modules.

All module kinds are listed: pure Python, built-in, frozen and extension
modules. Modules which are not available on some platforms and modules
disabled at Python build are also listed.

For packages, only sub-packages are listed, not sub-modules. For example,
``concurrent.futures`` is listed, but not ``concurrent.futures.base``.

Some special stdlib modules are excluded, like test and private modules.

It is a superset of the :attr:`sys.builtin_module_names` list.
```

**Comment by vstinner** [comment]
There is the dump of sys.module_names, 296 modules: http://paste.alacon.org/47013

**Comment by vstinner** [comment]
I modified the script to not ignore any module: with such hack, sys.module_names contains 1869 names. List of the 1575 ignored modules: http://paste.alacon.org/47014

I don't think that we should include these test modules and sub-modules. IMO only listing parent packages is enough. It's easy to detect that "asyncio" is a stdlib module from the "asyncio.base_events" name.

**Comment by vstinner** [comment]
I rebased my PR, squashed commits, and fixed a few comments of Tools/scripts/generate_module_names.py.

**Comment by vstinner** [comment]
"make regen-module-names" should be tested on macOS and FreeBSD, I'm not sure that setup.py reports properly missing modules in all cases.

**Comment by vstinner** [comment]
On FreeBSD, ``make regen-module-names`` does not change ``Python/module_names.h`` (``sys.module_names`` also contains 296 modules).

I wrote a short script to check which modules can be imported or not. Only 9 modules cannot be import on my FreeBSD VM:

```
_msi: False
_tkinter: False
msilib: False
msvcrt: False
spwd: False
tkinter: False
turtle: False
winreg: False
winsound: False
```

There are 5 modules specific to Windows (_msi, msilib, msvcrt, winreg, winsound), spwd doesn't exist on FreeBSD, _tkinter probably needs a missing build dependency (and turtle needs it).

import_all.py script:

```
import sys
import io

def import_ok(name):
    try:
        __import__(name)
    except ImportError:
        return False
    else:
        return True

stdout = sys.stdout
stderr = sys.stderr

for name in sys.module_names:
    sys.stdout = io.StringIO()
    sys.stderr = io.StringIO()
    ok = import_ok(name)
    sys.stdout = stdout
    sys.stderr = stderr
    print(f"{name}: {ok}")
```

**Comment by vstinner** [comment]
On Linux (on my Fedora 33 laptop), only 5 modules of sys.module_names cannot be imported, the 5 Windows specific modules:

```
_msi: False
msilib: False
msvcrt: False
winreg: False
winsound: False
```

**Comment by vstinner** [comment]
On Windows, 22 modules cannot be imported:

```
_crypt: False
_curses: False
_curses_panel: False
_dbm: False
_gdbm: False
_posixshmem: False
_posixsubprocess: False
crypt: False
curses: False
fcntl: False
grp: False
nis: False
ossaudiodev: False
posix: False
pty: False
pwd: False
readline: False
resource: False
spwd: False
syslog: False
termios: False
tty: False
```

Oh, there are 3 built-in modules on Windows which are not listed on Linux:

```
_winapi
_xxsubinterpreters
nt
```

I chose to exclude _xxsubinterpreters in Tools/scripts/generate_module_names.py:

```
    # Experimental module
    '_xxsubinterpreters',
```

I would prefer to have the same list on Linux and Windows.

**Comment by vstinner** [comment]
I updated the PR to add 3 modules (_winapi, _xxsubinterpreters, nt). sys.module_names now contains 299 modules on all platforms.

Note: I checked that all built-in modules listed in Modules/config.c on Linux and PC/config.c on Windwos are listed by Python/module_names.h.

**Comment by vstinner** [comment]
I created PR #24254 for bpo-42923 to only dump third party extensions on a Python fatal error (Py_FatalError(), faulthandler fatal signal). The PR is based on this PR.

**Comment by vstinner** [comment]
cc @ronaldoussoren

**Comment by vstinner** [comment]
@ronaldoussoren: The important part of this PR is the documentation. Do you think that it clearly describe what can found in the list and limitations? Or do you disagree with including modules which are not available?

**Comment by vstinner** [comment]
I rebased this PR which made it way simpler to only focus on adding the sys.module_names list.

I already merged the uncontroversial part in a private API. I started with a private list _Py_module_names in a new Python/module_names.h file: https://github.com/python/cpython/commit/cad8020cb83ec6d904f874c0e4f599e651022196 It unblocked bpo-42923 to dump third party extension modules at a fatal error.

**Comment by vstinner** [comment]
I enhanced Tools/scripts/generate_module_names.py to reorder the list and to avoid duplicates. sysmodule.c no longer has to remove duplicates at runtime.

**Comment by vstinner** [comment]
> I enhanced Tools/scripts/generate_module_names.py to reorder the list and to avoid duplicates. sysmodule.c no longer has to remove duplicates at runtime.

Hum, the generated list can be sorted as well. I simplified the runtime construction of sys.module_names even more.

File: Doc/library/sys.rst [hunk 1]
```diff
@@ -1060,6 +1062,31 @@ always available.
...
+   The list is the same on all platforms. Modules which are not available on
+   some platforms and modules disabled at Python build are also listed.
+   All module kinds are listed: pure Python, built-in, frozen and extension
+   modules.
+
+   For packages, only sub-packages are listed, not sub-modules. For example,
+   ``concurrent`` package and ``concurrent.futures`` sub-package are listed,
+   but not ``concurrent.futures.base`` sub-module.
+
+   Some special stdlib modules are excluded, like test and private modules.
```
Comment (pablogsal) [inline Doc/library/sys.rst]: Excluding tests is fine but why excluding private modules? They will appear in `sys.modules` and they belong to the standard library.

File: Doc/library/sys.rst [hunk 2]
```diff
@@ -1060,6 +1062,31 @@ always available.
...
+
+   See also the :attr:`sys.builtin_module_names` list.
+
+   .. note::
+      If a third party module has the same name than a standard library module
```
Comment (pablogsal) [inline Doc/library/sys.rst]: Why this note? How is this information relevant to this API? The name itself is the same, right? Or what we want to say is that there may be 3rd party modules with the same name as stdlib modules?

File: Python/pylifecycle.c [hunk 3]
```diff
@@ -2515,32 +2513,42 @@ _Py_DumpExtensionModules(int fd, PyInterpreterState *interp)
...
-        for (Py_ssize_t i=0; i < names_len; i++) {
-            const char *name = _Py_module_names[i];
-            if (PyUnicode_CompareWithASCIIString(key, name) == 0) {
-                is_stdlib_mod = 1;
-                break;
-            }
-        }
-        if (is_stdlib_mod) {
-            // Ignore stdlib extension module.
+        if (module_names != NULL && PySequence_Contains(module_names, key)) {
```
Comment (pablogsal) [inline Python/pylifecycle.c]: `PySequence_Contains` can raise an exception, no?

File: Doc/library/sys.rst [hunk 4]
```diff
@@ -1060,6 +1062,31 @@ always available.
         This is still called as a fallback if a :data:`meta_path` entry doesn't
         have a :meth:`~importlib.abc.MetaPathFinder.find_spec` method.
 
+.. data:: module_names
```
Comment (pablogsal) [inline Doc/library/sys.rst:1065]: Question: as the main use of this is checking if something is a stdlib module we can return a frozenset here. Any reason the order is important?

File: Doc/library/sys.rst [hunk 5]
```diff
@@ -1060,6 +1062,31 @@ always available.
...
+
+   A tuple of strings giving the names of standard library modules.
```
Comment (pablogsal) [inline Doc/library/sys.rst]:
```suggestion
   A tuple of strings containing the names of standard library modules.
```

File: Doc/library/sys.rst [hunk 6]
```diff
@@ -1060,6 +1062,31 @@ always available.
...
+
+   The list is the same on all platforms. Modules which are not available on
```
Comment (pablogsal) [inline Doc/library/sys.rst]:
```suggestion
   The output is the same on all platforms. Modules which are not available on
```

"the list" can be confused with the return type (although we say before is a tuple).

File: Doc/whatsnew/3.10.rst [hunk 7]
```diff
@@ -354,6 +354,10 @@ Add :data:`sys.orig_argv` attribute: the list of the original command line
 arguments passed to the Python executable.
 (Contributed by Victor Stinner in :issue:`23427`.)
 
+Add :data:`sys.module_names` attribute: the list of the standard library module
+names.
+(Contributed by Victor Stinner in :issue:`42955`.)
```
Comment (pablogsal) [inline Doc/whatsnew/3.10.rst]:
```suggestion
Add :data:`sys.module_names`, containing the list of the standard library module
names.
(Contributed by Victor Stinner in :issue:`42955`.)
```

File: Python/pylifecycle.c [hunk 8]
```diff
@@ -2515,32 +2513,42 @@ _Py_DumpExtensionModules(int fd, PyInterpreterState *interp)
...
+    if (module_names != NULL && !PyTuple_Check(module_names)) {
+        module_names = NULL;
+    }
+
+    // List extensions
     int header = 1;
     Py_ssize_t count = 0;
-    Py_ssize_t pos = 0;
-    PyObject *key, *value;
+    pos = 0;
```
Comment (pablogsal) [inline Python/pylifecycle.c:2540]: Why resetting `pos`? Is otherwise unused in the whole function for its value.

File: Doc/library/sys.rst (hunk 1 above)
Comment (vstinner) [reply Doc/library/sys.rst]: Well, my rationale is to keep the list looking nice for people who displays the list. There is no technical reason for hiding modules like _collections_abc or _pydecimal. I can easily shrink the IGNORE list in Tools/scripts/generate_module_names.py.

File: Doc/library/sys.rst (hunk 2 above)
Comment (vstinner) [reply Doc/library/sys.rst]: If you put an "asyncio.py" in the current directory, run Python and type "import asyncio", you get your module instead of the stdlib module. That's what I mean by "same name".

It's a note related to https://bugs.python.org/issue42955#msg385255 remark.

File: Python/pylifecycle.c (hunk 3 above)
Comment (vstinner) [reply Python/pylifecycle.c]: Since module_names can only be a tuple at this point, the code calls tuplecontains():

```
    for (i = 0, cmp = 0 ; cmp == 0 && i < Py_SIZE(a); ++i)
        cmp = PyObject_RichCompareBool(PyTuple_GET_ITEM(a, i), el, Py_EQ);
```

Well, I should inline the code and call PyUnicode_Compare() directly, to prevent any bad surprises with sys.module_names which contains something different than a string.

File: Doc/library/sys.rst (hunk 4 above)
Comment (vstinner) [reply Doc/library/sys.rst:1065]: I'm used to tuple, I have no preference between tuple and frozenset. It can be a frozenset.

File: Doc/library/sys.rst (hunk 5 above)
Comment (vstinner) [reply Doc/library/sys.rst]: The text is copied from sys.builtin_module_names. I'm fine with changing it, but I would also suggest to change sys.builtin_module_names documentation in that case.

File: Doc/library/sys.rst (hunk 1 above)
Comment (pablogsal) [reply Doc/library/sys.rst]: > Well, my rationale is to keep the list looking nice for people who displays the list.

But then is nor super usefull for people that want to use the list for reporting for example. Imagine you want to report 3rd party modules. You will iterate over `sys.modules` and check if the names are in this list. If we don't include private modules then the reporting will be wrong.

File: Doc/library/sys.rst (hunk 2 above)
Comment (pablogsal) [reply Doc/library/sys.rst]: I see, but how that affects this list? We should maybe rephrase saying "there is no guarantee that an imported module with a name in this collection is actually an stdlib module as any 3rd party module can iterpose itself by ...."

File: Python/pylifecycle.c (hunk 3 above)
Comment (pablogsal) [reply Python/pylifecycle.c]: > Well, I should inline the code and call PyUnicode_Compare() directly, to prevent any bad surprises with sys.module_names which contains something different than a string.

👍

File: Doc/library/sys.rst (hunk 4 above)
Comment (pablogsal) [reply Doc/library/sys.rst:1065]: I would prefer a frozenset, just because the main use is comparison. We can leave the actual type not specify and say that is an "iterable"

File: Doc/library/sys.rst (hunk 5 above)
Comment (pablogsal) [reply Doc/library/sys.rst]: As you prefer

**Comment by vstinner** [comment]
I updated the PR to take @pablogsal review in account:

* Change sys.module_names type to frozenset.
* No longer ignore private modules (only ignore test modules).
* Avoid unsafe PySequence_Contains().
* Rephrase the documentation.
* Remove the confusing note in the doc about sys.path and import.

Sorry, I amended my commit to be able to modify the commit message.

**Comment by vstinner** [comment]
@ronaldoussoren @serhiy-storchaka @pablogsal: I plan to merge this PR next monday. Please tell me if you want to review it before that.

When I created https://bugs.python.org/issue42955 I wasn't sure if sys.module_names would be useful, but then I found tons of use cases, and multiple persons told me that they need it for their projects (see the issue, I listed all of them).

Maybe we could add in addition a way to get paths of the stdlib, but I suggest to do that separately. Multiple use cases cannot import modules, but need to check the module name.
